java/core/src/java/org/apache/orc/impl/StringHashTableDictionary.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.orc.impl;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import org.apache.hadoop.io.Text;
+
+
+/**
+ * Using HashTable to represent a dictionary. The strings are stored as UTF-8 bytes
+ * and an offset for each entry. It is using chaining for collision resolution.
+ *
+ * This implementation is not thread-safe. It also assumes there's no reduction in the size of hash-table
+ * as it shouldn't happen in the use cases for this class.
+ */
+public class StringHashTableDictionary implements Dictionary {
+
+  // containing all keys every seen in bytes.
+  private final DynamicByteArray byteArray = new DynamicByteArray();
+  // containing starting offset of the key (in byte) in the byte array.
+  private final DynamicIntArray keyOffsets;
+
+  private final Text newKey = new Text();
+
+  private DynamicIntArray[] hashBuckets;
+
+  private int capacity;
+
+  private int threshold;
+
+  private float loadFactor;
+
+  private static float DEFAULT_LOAD_FACTOR = 0.75f;
+
+  private static final int BUCKET_SIZE = 20;
```

Comment:
How did we come up with 20 here?

Reply:
It is just randomly picked ...

Reply:
Was just thinking that for the default batches of  size 1024, we use Dict encoding if less than 20% of are distinct we would need around 0.6 - 0.7 of the batch size?
In that sense bucketSize of 5 or 6 could be enough right?

Might be missing something but lets add a comment here behind the logic

Reply:
Can you elaborate a bit more on how 5/6 is deduced?  the particular point I am confusing is its relation to the batch size, since dictionary size is more associated with stripe size right ?

Reply:
My rough estimate: 
Given default stripe size 64MB, assuming each row were around 500B, 

`4096 * 0.75 (capacity without resize) * avgBucketSize * 5 (20% distinct) = 64 * 1024 * 1024 / 500`

avgBucketSize ~8 in this case. Let me know if it makes sense/converge with your thoughts

Reply:
Hey @autumnust  you are right, it is more associated with the stripe size so your estimation makes sense. Thanks for clarifying!